### DISCUSSION TITLE
Questions regarding additional structs

### DISCUSSION THREAD

**Discussion by drauch** [microsoft/CsWin32#1002] [Q&A]
Really like the concept of the library and started to use it today.

I've used it to generate `SignerSignEx2`, this generates lots of additional structs, and I've problems filling them:

1) strings in those structs are PCWSTR and not System.String typed. How to properly convert a System.String into a PCWSTR? Wouldn't it be nice if those structs would contain System.String properties? :-)

2) How to properly fill the cbSize of those structs? Is this something that is done automatically in the ctor? It doesn't look like it, so I have to use `(uint) Marshal.SizeOf<STRUCT_TYPE>()` am I right?

3) Unions are strangely called "Anonymous", is this correct or a generation error?

Best regards,
D.R.

**Comment by drauch** [comment]
Another question:

`CreateFile` returns a `SafeHandle`, the struct `SIGNER_FILE_INFO` wants a `HANDLE`. How to do that? Is `(HANDLE) file.DangerousGetHandle()` the way to go?

**Reply by AArnott** [reply]
Yes, that's a fine approach.

**Comment by drauch** [comment]
One more:

`CertOpenStore` has a parameter of `HCRYPTPROV` which should be set to null. I can't pass `null` nor `IntPtr.Zero` though. How to pass NULL as this paramteter? Is `new HCRYPTPROV_LEGACY(nuint.Zero)` the way to go?

All the best,
D.R.

**Reply by AArnott** [reply]
Yes, that works. But you can also just pass in `default`, which amounts to the same thing.

**Comment by drauch** [comment]
Also generating constants defined in https://learn.microsoft.com/en-us/windows/win32/seccrypto/alg-id does not work, how can they be found? Do I have to manually create those?

Best regards,
D.R.

PS: Sorry for the spam :-)

**Reply by AArnott** [reply]
Those aren't in the metadata, probably because they aren't technically constants in header files but rather macros. @mikebattista what are the odds these can be scraped?

**Reply by mikebattista** [reply]
We can manually define those.

**Reply by drauch** [reply]
Can you tell me when those will be available in a beta NuGet package approx.? :-) 

Best regards,
D.R.

**Comment by AArnott** [comment]
> strings in those structs are PCWSTR and not System.String typed. How to properly convert a System.String into a PCWSTR? Wouldn't it be nice if those structs would contain System.String properties? :-)

Yes, it would be nice in some cases. But it converts a struct into what the CLR calls a managed type, so you can no longer get a pointer to it, as is required by many APIs. So while we offer `string` APIs on method parameters, we can't offer them in struct fields. 
But we do offer some helpers. `PCWSTR.ToString()` for example is implemented. And if you have a string you can create a PCWSTR and initialize it using a few techniques. A good discussion on that can be seen or continued here: https://github.com/microsoft/CsWin32/discussions/912#discussioncomment-5580291

> How to properly fill the cbSize of those structs? Is this something that is done automatically in the ctor? It doesn't look like it, so I have to use (uint) Marshal.SizeOf<STRUCT_TYPE>() am I right?

Yes, you have to initialize those fields yourself. For a few uncommon variable length structs, you'd have to do even more. 

> Unions are strangely called "Anonymous", is this correct or a generation error?

The names come from the metadata. We propose to improve this as tracked by #408.

**Comment by mikebattista** [comment]
Ideally CsWin32 would autopopulate cbSize fields via https://github.com/microsoft/CsWin32/issues/108.

**Comment by drauch** [comment]
> Yes, it would be nice in some cases. But it converts a struct into what the CLR calls a managed type, so you can no longer get a pointer to it, as is required by many APIs. So while we offer string APIs on method parameters, we can't offer them in struct fields.
But we do offer some helpers. PCWSTR.ToString() for example is implemented. And if you have a string you can create a PCWSTR and initialize it using a few techniques. A good discussion on that can be seen or continued here: https://github.com/microsoft/CsWin32/discussions/912#discussioncomment-5580291

The main annoyance isn't string for me, it's sub structs:

```
// I'd like to write code like that:
var myStruct = CreateStruct();
PInvoke.MyMethod(..., myStruct, ...);
[...]
OUTER_STRUCT CreateStruct ()
{
    INNER_STRUCT is;
    // fill is

    OUTER_STRUCT os;
    // fill os
    os.Inner = &is; // unfortunately pointer type instead of a c# reference type

    return os;
}
```

That doesn't work though, as the address `&is` is invalid once I return from `CreateStruct`. Now I have to write code like that:

```
var disposers = new List<Action>();
try
{
    var myStruct = CreateStruct(disposers);
    PInvoke.MyMethod(..., myStruct, ...);
}
finally
{
    foreach(var disposer in disposers)
        disposer();
}

[...]
OUTER_STRUCT CreateStruct (List<Action> disposers)
{
    INNER_STRUCT is;
    // fill is
    INNER_STRUCT* isPinned = GCHandle.Alloc(is, GCHandleType.Pinned);
    disposers.Add(() => isPinned.Free());

    OUTER_STRUCT os;
    // fill os
    os.Inner = (INNER_STRUCT*) is.AddrOfPinnedObject();

    return os;
}
```

As you can see this is much more complex now. If you have a better workaround I'd be glad :-)

Best regards,
D.R.

**Reply by AArnott** [reply]
That's interesting. I suppose a `[MarshalAs(UnmanagedType.LPStruct)]` applied to the `Inner` field could make this simpler, in the allowMarshaling case. But it would make the outer struct itself unpinnable, which could lead to cascading issues. It's worth investigating though. Wanna file an issue requesting this specifically?

**Reply by drauch** [reply]
Sure => https://github.com/microsoft/CsWin32/issues/1003

**Comment by drauch** [comment]
@AArnott & @mikebattista  : Many thanks for the quick replies - and this great gem of a library ❤️
